tests/test_controlled_gates.py

```diff
@@ -0,0 +1,129 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import pytest
+import cmath as cm
+
+from perceval import BasicState, Port, Processor, Encoding, PostSelect
+from perceval.algorithm import Analyzer
+from perceval.utils import generate_all_logical_states
+from perceval.simulators import SimulatorFactory
+from perceval.components import catalog, BS, get_basic_state_from_ports
+
+
+def check_controlled_gates_and_get_performance(n, alpha, processor, herald_states, error=1E-6):
+    """Check if the CC..Z(alpha) is correct
+
+    Meaning checking that the CC...Z(alpha) gate probability amplitude matrix should be: 
+                        Id_(2^n) +(exp(i.alpha)-1) |1...1><1...1|
+
+    :param processor: CCZ processor to check, we assume that ctrl is on [0,1] and [2,3]
+    and data on [4,5]
+    :param herald_states: Basic state corresponding to the heralded or ancillary modes
+    at the end of the circuit
+    :param error: Tolerance for the modulus of the prob_amplitude (when not expecting 0), defaults to 1E-6
+    This param represent whether this CCZ gate is balanced
+    """
+    ports = [Port(Encoding.DUAL_RAIL, "")] * n
+    states = [get_basic_state_from_ports(ports, state) * herald_states for state in generate_all_logical_states(n)]
+    sim = SimulatorFactory().build(processor)
+
+    data_state = BasicState("|0,1"+(n-1)*",0,1"+">") * herald_states
+    modulus_value = None
+    phase_value = None
+    for i_state in states:
+        for o_state in states:
+            pa = sim.prob_amplitude(i_state, o_state)
+            modulus = abs(pa)
+            phase = cm.phase(pa)
+            if i_state == o_state:
+                if modulus_value is None:
+                    modulus_value = modulus
+                assert pytest.approx(modulus, error) == modulus_value
+                assert modulus != 0
+
+                if i_state != data_state:
+                    if phase_value is None:
+                        phase_value = phase
+                    assert pytest.approx(phase) == phase_value % ( 2 * cm.pi)
+                else:
+                    delta_phase = (phase - (phase_value + alpha)+ cm.pi) % (2*cm.pi) - cm.pi
+                    assert pytest.approx(delta_phase) == 0 
+
+            else:
+                assert pytest.approx(modulus) == 0
+    return modulus_value
+
+
+def test_controlled_gates():
+    for n in [2, 3, 4]:
+        for alpha in [cm.pi, cm.pi/3, -cm.pi/4, 1.]:
+
+            # Testing phases and modulus of CCZ
+            modulus_gate = check_controlled_gates_and_get_performance(n, alpha,
+                catalog['postprocessed controlled gate'].build_processor(n = n, alpha = alpha ), BasicState("|0"+(2*n-1)*",0"+">"))
+
+@pytest.mark.skip(reason="redundant with overhead test")
+def test_inverted_sub_cnot():
```

Comment:
Why do you add this test in this file ? You can let it tho if you delete the file 
test_3qbits_gates.py